src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala

```diff
@@ -0,0 +1,162 @@
+      val stdout  = mutable.ListBuffer[String]()
+      val sink    = new AsyncStreamSink(process.getInputStream, s => stdout.append(s))
+      // True if the command can execute with no error
+      val canExecute = process.waitFor() == 0
+      sink.close()
+
+      (canExecute, stdout.toList.mkString)
+    }
+    catch {case e: Exception => (false, "")}
+  }
+
+  /** Returns true if the tool is available and false otherwise. */
+  lazy val Available: Boolean = {
+    execArgs(Executable +: TestCommand: _*)._1
+  }
+}
+
+trait CanRunScript {
+  self: CommandLineTool =>
+  /** Suffix for scripts that can be run by this tool. */
+  val Suffix: String
+
+  /** Executes a script from the classpath if the tested executable is available. */
+  def execIfAvailable(scriptResource: String, args: String*): Unit =
+    try {
+      if (Available) exec(scriptResource, args:_*)
+    } catch {case e: Exception => throw new Exception(s"Cannot execute script from $scriptResource with args:$args") }
+
+  /** Executes from a script stored at a Path if the tested executable is available. */
+  def execIfAvailable(script: Path, args: String*): Unit =
+    try {
+      if (Available) exec(script, args:_*)
+    } catch {case e: Exception => throw new Exception(s"Cannot execute script from $script with args: $args") }
+
+  /** Executes a script from the classpath. */
+  def exec(scriptResource: String, args: String*): Unit =
+    Try { writeResourceToTempFile(scriptResource) }.map(path => exec(path, args:_*))
+
+  /** Executes from a script stored at a Path. */
+  def exec(script: Path, args: String*): Unit = try {
+    val command = Executable +: script.toAbsolutePath.toString +: args
+    val process = new ProcessBuilder(command:_*).redirectErrorStream(false).start()
+    val pipe1   = Io.pipeStream(process.getErrorStream, logger.info)
+    val pipe2   = Io.pipeStream(process.getInputStream, logger.debug)
+    val retval  = process.waitFor()
+    pipe1.close()
+    pipe2.close()
+
+    if (retval != 0) throw ToolException(retval)
+  } catch {case e: Exception => throw new Exception(s"Cannot execute script from path $script with args: $args") }
+
+  /** Extracts a resource from the classpath and writes it to a temp file on disk. */
+  private def writeResourceToTempFile(resource: String): Path = {
+    val lines = Io.readLinesFromResource(resource).toSeq
+    val path  = Io.makeTempFile("script.", suffix = Suffix)
+    path.toFile.deleteOnExit()
+    Io.writeLines(path, lines)
+    path
+  }
+}
+
+trait Versioned {
+  self: CommandLineTool =>
+  val VersionFlag: String      = "--version"
+  val TestCommand: Seq[String] = Seq(VersionFlag)
+
+  /** Returns version of the tool */
+  lazy val Version: String = execArgs(Executable +: TestCommand:_*)._2
+}
+
+
+trait Modular {
+  self: CommandLineTool =>
+
+  /** Command used to test if a module is included with the tested executable. */
+  def TestModuleCommand(module: String): Seq[String]
```

Comment:
```suggestion
  def TestModuleArgs(module: String): Seq[String]
```

Here and throughout?